docs/timeline.md

```diff
@@ -34,9 +31,35 @@ erDiagram
     }
 
     audience 1--0+ audience_member : has
-    actor 1--0+ audience_member : is
-    profile 1--1+ actor : controls
-    share 1--1+ audience : with
-    actor 1--0+ share : initiates
-    share 1+--1 object : refers_to
-```
\ No newline at end of file
+    profile 1--0+ audience_member : is
+    feeds 1--1+ audience : with
+    profile 1--0+ feeds : initiates
+    feeds 1+--1 object : refers_to
+```
+
+## Background
+
+Timelines are the central interaction mechanism for a microblogging service. They need to service extremely high queries rates for relatively large data volumes, as well as constant writes. The returned results need to respect the followings, filters, and lists that are specified for the profile, as well as the creator's and instance's blocks/mutes/etc. As far as I can tell Mastodon (and Calkey/Firefish) accomplishes this by just dumping regular posts into redis and then issuing fairly complex queries to filter down to the correct set of posts to fill out a user's timeline. We're not doing that. Redis is expensive, and scaling it is even more expensive. Also, maybe it's just me, but I think just loading everything into RAM is actually not great system design? And even with the performance of entirely in-memory storage, querying Redis isn't free. Thanks to the hex pattern, changing course and using Redis in the future remains an option, of course.
+
+Instead, we'll be using a timeseries database. Specifically, TimescaleDB, at least to start. Timeseries DBs are analytic—as opposed to transactional—databases. The basic premise is that by giving up on relational integrity, the database can very quickly select, filter, and aggregate very large sets of records. And timeseries DBs in particular, are designed around doing this with data that is ordered and queried chronologically. Exactly like a timeline.
```

Comment:
What are the operational impacts of this for admins? I really like this choice, but people may be more familiar with operating relational SQL databases vs a TSDB.

Reply:
Timescale is actually a plugin for Postgres, so the operational concern is installing that plugin. Or providing connection info for a managed instance. I plan to put feeds in its own database. It can be on the same postgres instance as everything else, or a different one. And it shouldn't be too difficult to migrate from a shared to dedicated instance when a server grows to that size.